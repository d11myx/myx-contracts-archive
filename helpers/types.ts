import { IPool } from '../types';
import { eNetwork } from './constants';
import { IFundingRate } from '../types/contracts/core/FundingRate';
import type { PromiseOrValue } from '../types/common';
import type { BigNumberish } from 'ethers';

export interface SymbolMap<T> {
    [symbol: string]: T;
}

export type ParamsPerNetwork<T> = {
    [k in eNetwork]?: T;
};

export type TradingFeeTier = {
    takerFee: PromiseOrValue<BigNumberish>;
    makerFee: PromiseOrValue<BigNumberish>;
};

export interface PairInfoConfig {
    pairTokenDecimals: number;
    useWrappedNativeToken: boolean;
    pair: IPool.PairStruct;
    tradingConfig: IPool.TradingConfigStruct;
    tradingFeeConfig: IPool.TradingFeeConfigStruct & TradingFeeTier;
    fundingFeeConfig: IFundingRate.FundingFeeConfigStruct;
}

export interface ReserveConfiguration {
    MarketTokenSymbol: string;
    MarketTokenName: string;
    MarketTokenDecimals: number;
    MarketTokenAddress: ParamsPerNetwork<string>;
    WrapperTokenAddress: ParamsPerNetwork<string>;
    PairsConfig: SymbolMap<PairInfoConfig>;
    PairAssets: ParamsPerNetwork<SymbolMap<string>>;
    ExecuteOrderTimeDelay: number;
    OraclePriceFeedAddress: ParamsPerNetwork<string>;
    OraclePriceId: ParamsPerNetwork<SymbolMap<string>>;
    UniswapRouterAddress: ParamsPerNetwork<string>;
    UniswapTokenPathConfig: ParamsPerNetwork<SymbolMap<string>>;
}
