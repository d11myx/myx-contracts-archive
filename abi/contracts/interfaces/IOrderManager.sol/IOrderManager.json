[
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "account",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "CancelDecreaseOrder",
    "type": "event"
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "account",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "CancelIncreaseOrder",
    "type": "event"
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "account",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      },
      {
        "indexed": false,
        "internalType": "int256",
        "name": "collateral",
        "type": "int256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "pairIndex",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "openPrice",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "sizeAmount",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "bool",
        "name": "isLong",
        "type": "bool"
      },
      {
        "indexed": false,
        "internalType": "bool",
        "name": "abovePrice",
        "type": "bool"
      }
    ],
    "name": "CreateDecreaseOrder",
    "type": "event"
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "account",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "pairIndex",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      },
      {
        "indexed": false,
        "internalType": "int256",
        "name": "collateral",
        "type": "int256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "openPrice",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "bool",
        "name": "isLong",
        "type": "bool"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "sizeAmount",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "tpPrice",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "tpAmount",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "slPrice",
        "type": "uint256"
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "slAmount",
        "type": "uint256"
      }
    ],
    "name": "CreateIncreaseOrder",
    "type": "event"
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "oldAddress",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "address",
        "name": "newAddress",
        "type": "address"
      }
    ],
    "name": "UpdatePositionManager",
    "type": "event"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "isIncrease",
            "type": "bool"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          }
        ],
        "internalType": "struct IOrderManager.PositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "name": "addOrderToPosition",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "account",
        "type": "address"
      },
      {
        "internalType": "uint256",
        "name": "pairIndex",
        "type": "uint256"
      },
      {
        "internalType": "bool",
        "name": "isLong",
        "type": "bool"
      }
    ],
    "name": "cancelAllPositionOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      },
      {
        "internalType": "bool",
        "name": "isIncrease",
        "type": "bool"
      }
    ],
    "name": "cancelOrder",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "int256",
            "name": "sizeAmount",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.CreateOrderRequest",
        "name": "request",
        "type": "tuple"
      }
    ],
    "name": "createOrder",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "decreaseLimitOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "decreaseMarketOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "getDecreaseOrder",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "triggerPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "abovePrice",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "needADL",
            "type": "bool"
          }
        ],
        "internalType": "struct TradingTypes.DecreasePositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "getIncreaseOrder",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.IncreasePositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "key",
        "type": "bytes32"
      }
    ],
    "name": "getPositionOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "isIncrease",
            "type": "bool"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          }
        ],
        "internalType": "struct IOrderManager.PositionOrder[]",
        "name": "",
        "type": "tuple[]"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "increaseLimitOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "increaseMarketOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "key",
        "type": "bytes32"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "positionHasTpSl",
    "outputs": [
      {
        "internalType": "bool",
        "name": "",
        "type": "bool"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      }
    ],
    "name": "removeDecreaseLimitOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      }
    ],
    "name": "removeDecreaseMarketOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      }
    ],
    "name": "removeIncreaseLimitOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      }
    ],
    "name": "removeIncreaseMarketOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "isIncrease",
            "type": "bool"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          }
        ],
        "internalType": "struct IOrderManager.PositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "name": "removeOrderFromPosition",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      },
      {
        "internalType": "bool",
        "name": "needADL",
        "type": "bool"
      }
    ],
    "name": "setOrderNeedADL",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "key",
        "type": "bytes32"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      },
      {
        "internalType": "bool",
        "name": "has",
        "type": "bool"
      }
    ],
    "name": "setPositionHasTpSl",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "newAddress",
        "type": "address"
      }
    ],
    "name": "updatePositionManager",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  }
]
