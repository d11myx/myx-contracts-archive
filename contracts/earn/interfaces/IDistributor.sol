// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IDistributor {
    function updateRoot(bytes32 _merkleRoot) external;
    function claimForAccount(address account, address receiver, uint256 amount, bytes32[] calldata merkleProof) external;
}
