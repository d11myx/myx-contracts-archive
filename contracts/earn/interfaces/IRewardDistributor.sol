// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IRewardDistributor {
    function updateRoot(bytes32 _merkleRoot, uint256 _amount) external;

    function claimForAccount(
        address account,
        address receiver,
        uint256 amount,
        bytes32[] calldata merkleProof
    ) external;
}
