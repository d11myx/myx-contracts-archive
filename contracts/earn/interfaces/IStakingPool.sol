// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IStakingPool {
    function stakeForAccount(address funder, address account, uint256 amount) external;
    function unstakeForAccount(address account, address receiver, uint256 amount) external;
}
