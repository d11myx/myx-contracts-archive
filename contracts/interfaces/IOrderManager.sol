// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import "../libraries/type/TradingTypes.sol";

interface IOrderManager {

    event UpdatePositionManager(address oldAddress, address newAddress);

    event CreateIncreaseOrder(
        address account,
        uint256 orderId,
        uint256 pairIndex,
        TradingTypes.TradeType tradeType,
        int256 collateral,
        uint256 openPrice,
        bool isLong,
        uint256 sizeAmount,
        uint256 tpPrice,
        uint256 tpAmount,
        uint256 slPrice,
        uint256 slAmount
    );

    event CreateDecreaseOrder(
        address account,
        uint256 orderId,
        TradingTypes.TradeType tradeType,
        int256 collateral,
        uint256 pairIndex,
        uint256 openPrice,
        uint256 sizeAmount,
        bool isLong,
        bool abovePrice
    );

    event CancelIncreaseOrder(address account, uint256 orderId, TradingTypes.TradeType tradeType);
    event CancelDecreaseOrder(address account, uint256 orderId, TradingTypes.TradeType tradeType);

    struct PositionOrder {
        address account;
        uint256 pairIndex;
        bool isLong;
        bool isIncrease;
        TradingTypes.TradeType tradeType;
        uint256 orderId;
        uint256 sizeAmount;
    }

    function increaseMarketOrdersIndex() external view returns(uint256);
    function decreaseMarketOrdersIndex() external view returns(uint256);
    function increaseLimitOrdersIndex() external view returns(uint256);
    function decreaseLimitOrdersIndex() external view returns(uint256);

    function positionHasTpSl(bytes32 key, TradingTypes.TradeType tradeType) external view returns(bool);

    function getPositionOrders(bytes32 key) external view returns(PositionOrder[] memory);

    function updatePositionManager(address newAddress) external;

    function createOrder(TradingTypes.CreateOrderRequest memory request) external returns (uint256 orderId);

    function cancelOrder(uint256 orderId, TradingTypes.TradeType tradeType, bool isIncrease) external;

    function getIncreaseOrder(uint256 orderId, TradingTypes.TradeType tradeType) external view returns (TradingTypes.IncreasePositionOrder memory order);
    function getDecreaseOrder(uint256 orderId, TradingTypes.TradeType tradeType) external view returns (TradingTypes.DecreasePositionOrder memory order);

    function addOrderToPosition(PositionOrder memory order) external;
    function removeOrderFromPosition(PositionOrder memory order) external;

    function setPositionHasTpSl(bytes32 key, TradingTypes.TradeType tradeType, bool has) external;

    function removeIncreaseMarketOrders(uint256 orderId) external;
    function removeIncreaseLimitOrders(uint256 orderId) external;
    function removeDecreaseMarketOrders(uint256 orderId) external;
    function removeDecreaseLimitOrders(uint256 orderId) external;

    function setOrderNeedADL(uint256 orderId, TradingTypes.TradeType tradeType, bool needADL) external;
}
