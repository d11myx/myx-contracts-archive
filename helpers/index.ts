export * from './constants';
export * from './contract-deployments';
export * from './contract-getters';
export * from './deploy-ids';
export * from './env';
export * from './hardhat-config-helpers';
export * from './init-helper';
export * from './market-config-helper';
export * from './types';
export * from './utilities/signer';
export * from './utilities/tx';
export * from '../tasks/misc/print-deployments';

import { loadTasks } from './hardhat-config-helpers';

const TASK_FOLDERS = ['../tasks/misc'];

// Load all plugin tasks
loadTasks(TASK_FOLDERS);
