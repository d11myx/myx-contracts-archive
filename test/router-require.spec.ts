import { ethers } from 'hardhat';
import { TestEnv, newTestEnv } from './helpers/make-suite';
import { mintAndApprove } from './helpers/misc';
import { MAX_UINT_AMOUNT, TradeType, deployMockCallback, waitForTx } from '../helpers';
import { IPool } from '../types';
import { expect } from './shared/expect';
import { TradingTypes } from '../types/contracts/interfaces/IOrderManager';

describe('Router: check require condition, trigger errors', async () => {
    const pairIndex = 0;
    let testEnv: TestEnv;

    before(async () => {
        testEnv = await newTestEnv();
        const {
            users: [depositor],
            usdt,
            btc,
            pool,
        } = testEnv;

        // add liquidity
        const indexAmount = ethers.utils.parseUnits('20000', 18);
        const stableAmount = ethers.utils.parseUnits('300000000', 18);
        let testCallBack = await deployMockCallback(btc.address, usdt.address);
        await mintAndApprove(testEnv, btc, indexAmount, depositor, testCallBack.address);
        await mintAndApprove(testEnv, usdt, stableAmount, depositor, testCallBack.address);
        await testCallBack.connect(depositor.signer).addLiquidity(pool.address, pairIndex, indexAmount, stableAmount);
    });
    after(async () => {});

    describe('createIncreaseOrder permission check', async () => {
        it('check msg.sender whith request.account', async () => {
            const {
                keeper,
                deployer,
                users: [user1, user2],
                usdt,
                router,
                executor,
                tradingVault,
                orderManager,
            } = testEnv;

            const amount = ethers.utils.parseUnits('10000', 18);
            const collateral = ethers.utils.parseUnits('10000', 18);
            const size = ethers.utils.parseUnits('10', 18);

            await waitForTx(await usdt.connect(deployer.signer).mint(user1.address, collateral));
            await usdt.connect(user1.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

            await waitForTx(await usdt.connect(deployer.signer).mint(user2.address, collateral));
            await usdt.connect(user2.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

            // setting: request.account = user1
            const increasePositionRequest: TradingTypes.CreateOrderRequestStruct = {
                account: user1.address,
                pairIndex: pairIndex,
                tradeType: TradeType.MARKET,
                collateral: collateral,
                openPrice: ethers.utils.parseUnits('30000', 30),
                isLong: true,
                sizeAmount: size,
                tp: 0,
                tpPrice: 0,
                sl: 0,
                slPrice: 0,
            };

            // setting createIncreateOrder: msg.sender = user
            const orderId = await orderManager.increaseMarketOrdersIndex();
            await router.connect(user2.signer).createIncreaseOrder(increasePositionRequest);
            await executor.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET);
            // await expect(router.connect(user2.signer).createIncreaseOrder(increasePositionRequest)).to.be.revertedWith('not order sender or handler');
        });

        // TODO: Function to be implemented
        // it('Check if the user <isFrozen>', async () => {
        //     const {
        //         keeper,
        //         deployer,
        //         users: [ user1, user2],
        //         usdt,
        //         tradingRouter,
        //         tradingVault,
        //         executeRouter
        //     } = testEnv;

        //     const collateral = ethers.utils.parseUnits('10000', 18);
        //     const size = ethers.utils.parseUnits('10', 18);

        //     await waitForTx(await usdt.connect(deployer.signer).mint(user1.address, collateral));
        //     await usdt.connect(user1.signer).approve(tradingRouter.address, MAX_UINT_AMOUNT);

        //     await waitForTx(await usdt.connect(deployer.signer).mint(user2.address, collateral));
        //     await usdt.connect(user2.signer).approve(tradingRouter.address, MAX_UINT_AMOUNT);

        //     const increasePositionRequest: TradingTypes.IncreasePositionRequestStruct = {
        //         account: user1.address,
        //         pairIndex: pairIndex,
        //         tradeType: TradeType.MARKET,
        //         collateral: collateral,
        //         openPrice: ethers.utils.parseUnits('30000', 30),
        //         isLong: true,
        //         sizeAmount: size,
        //         tp: 0,
        //         tpPrice: 0,
        //         sl: 0,
        //         slPrice: 0
        //     };

        //     const isFrozen = tradingVault.isFrozen(user1.address);

        //     const orderId = await tradingRouter.increaseMarketOrdersIndex();
        //     await expect(router.connect(user1.signer).createIncreaseOrder(increasePositionRequest)).to.be.reverted;
        // });

        describe('disable pair', async () => {
            after('afer enable pair', async () => {
                const { pool } = testEnv;

                const pair = await pool.getPair(pairIndex);
                console.log(`pair: `, pair);
                const newPair: IPool.PairStruct = {
                    indexToken: pair.indexToken,
                    stableToken: pair.stableToken,
                    pairToken: pair.pairToken,
                    enable: true,
                    kOfSwap: pair.kOfSwap,
                    expectIndexTokenP: pair.expectIndexTokenP,
                    addLpFeeP: pair.addLpFeeP,
                };
                await pool.updatePair(pairIndex, newPair);
            });

            it('pair is enable', async () => {
                const {
                    keeper,
                    deployer,
                    users: [trader],
                    usdt,
                    router,
                    pool,
                    orderManager,
                } = testEnv;

                // disable pair
                const pairBef = await pool.getPair(pairIndex);
                console.log(`pair: `, pairBef);
                const newPair: IPool.PairStruct = {
                    indexToken: pairBef.indexToken,
                    stableToken: pairBef.stableToken,
                    pairToken: pairBef.pairToken,
                    enable: false,
                    kOfSwap: pairBef.kOfSwap,
                    expectIndexTokenP: pairBef.expectIndexTokenP,
                    addLpFeeP: pairBef.addLpFeeP,
                };
                await pool.updatePair(pairIndex, newPair);

                const pairAft = await pool.getPair(pairIndex);
                console.log(`pairAft: `, pairAft);

                const collateral = ethers.utils.parseUnits('10000', 18);
                const size = ethers.utils.parseUnits('10', 18);

                await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, collateral));
                await usdt.connect(trader.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

                const increasePositionRequest: TradingTypes.CreateOrderRequestStruct = {
                    account: trader.address,
                    pairIndex: pairIndex,
                    tradeType: TradeType.MARKET,
                    collateral: collateral,
                    openPrice: ethers.utils.parseUnits('30000', 30),
                    isLong: true,
                    sizeAmount: size,
                    tp: 0,
                    tpPrice: 0,
                    sl: 0,
                    slPrice: 0,
                };

                const orderId = await orderManager.increaseMarketOrdersIndex();
                await expect(
                    router.connect(trader.signer).createIncreaseOrder(increasePositionRequest),
                ).to.be.revertedWith('trade pair not supported');
                // await executer.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET);
            });
        });

        describe('check increase sizeAmount', async () => {
            it('sizeAmount = 0, trigger error: size eq 0', async () => {
                const {
                    keeper,
                    deployer,
                    users: [trader],
                    usdt,
                    router,
                    orderManager,
                } = testEnv;

                const collateral = ethers.utils.parseUnits('10000', 18);

                await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, collateral));
                await usdt.connect(trader.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

                const increasePositionRequest: TradingTypes.CreateOrderRequestStruct = {
                    account: trader.address,
                    pairIndex: pairIndex,
                    tradeType: TradeType.MARKET,
                    collateral: collateral,
                    openPrice: ethers.utils.parseUnits('30000', 30),
                    isLong: true,
                    sizeAmount: 0,
                    tp: 0,
                    tpPrice: 0,
                    sl: 0,
                    slPrice: 0,
                };

                const orderId = await orderManager.increaseMarketOrdersIndex();
                await expect(
                    router.connect(trader.signer).createIncreaseOrder(increasePositionRequest),
                ).to.be.revertedWith('size eq 0');
                // await executeRouter.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET)
            });

            it('sizeAmount < tradingConfig.minTradeAmount, trigger error: invalid trade size', async () => {
                const {
                    keeper,
                    deployer,
                    users: [trader],
                    usdt,
                    router,
                    pool,
                    orderManager,
                } = testEnv;

                const tradingConfig = await pool.getTradingConfig(pairIndex);
                const minTradeAmount = tradingConfig.minTradeAmount;
                const maxTradeAmount = tradingConfig.maxTradeAmount;

                console.log(`--minTradeAmount: `, minTradeAmount);
                console.log(`--maxTradeAmount: `, maxTradeAmount);

                const collateral = ethers.utils.parseUnits('10000', 18);
                const sizeAmount = ethers.utils.parseUnits('5', 15);

                await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, collateral));
                await usdt.connect(trader.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

                const increasePositionRequest: TradingTypes.CreateOrderRequestStruct = {
                    account: trader.address,
                    pairIndex: pairIndex,
                    tradeType: TradeType.MARKET,
                    collateral: collateral,
                    openPrice: ethers.utils.parseUnits('30000', 30),
                    isLong: true,
                    sizeAmount: sizeAmount,
                    tp: 0,
                    tpPrice: 0,
                    sl: 0,
                    slPrice: 0,
                };

                const orderId = await orderManager.increaseMarketOrdersIndex();
                await expect(
                    router.connect(trader.signer).createIncreaseOrder(increasePositionRequest),
                ).to.be.revertedWith('invalid trade size');
                // await executer.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET);
            });

            it('sizeAmount > tradingConfig.maxTradeAmount, trigger error: invalid trade size', async () => {
                const {
                    keeper,
                    deployer,
                    users: [trader],
                    usdt,
                    pool,
                    router,
                    orderManager,
                } = testEnv;

                const tradingConfig = await pool.getTradingConfig(pairIndex);
                const minTradeAmount = tradingConfig.minTradeAmount;
                const maxTradeAmount = tradingConfig.maxTradeAmount;

                console.log(`minTradeAmount: `, minTradeAmount);
                console.log(`maxTradeAmount: `, maxTradeAmount);

                const amount = ethers.utils.parseUnits('30000', 18);
                const collateral = ethers.utils.parseUnits('1000', 18);
                const sizeAmount = ethers.utils.parseUnits('100001', 18);

                await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, amount));
                await usdt.connect(trader.signer).approve(orderManager.address, MAX_UINT_AMOUNT);

                const increasePositionRequest: TradingTypes.CreateOrderRequestStruct = {
                    account: trader.address,
                    pairIndex: pairIndex,
                    tradeType: TradeType.MARKET,
                    collateral: collateral,
                    openPrice: ethers.utils.parseUnits('30000', 30),
                    isLong: true,
                    sizeAmount: sizeAmount,
                    tp: 0,
                    tpPrice: 0,
                    sl: 0,
                    slPrice: 0,
                };

                const orderId = await orderManager.increaseMarketOrdersIndex();
                await expect(
                    router.connect(trader.signer).createIncreaseOrder(increasePositionRequest),
                ).to.be.revertedWith('invalid trade size');
                // await executer.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET);
            });
        });
    });
});
