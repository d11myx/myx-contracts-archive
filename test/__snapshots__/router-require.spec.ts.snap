// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router: check require condition, trigger errors createIncreaseOrder permission check check increase sizeAmount sizeAmount > tradingConfig.maxTradeAmount, trigger error: invalid trade size 1`] = `305520`;
